File: backend/validation.py
"""Regras compartilhadas com o formulário JavaScript."""
import json
import re
from pathlib import Path
from uuid import UUID

RULES = json.loads((Path(__file__).resolve().parents[1] / "shared/contact-rules.json").read_text(encoding="utf-8"))


def validate_contact(payload):
    if not isinstance(payload, dict):
        return {}, {"form": ["Envie um objeto JSON."]}
    cleaned, errors = {}, {}
    for name, rule in RULES["fields"].items():
        value = payload.get(name)
        if not isinstance(value, str):
            errors[name] = [rule["message"]]
            continue
        value = value.strip()
        cleaned[name] = value
        if len(value) < rule["min"]:
            errors[name] = [rule["message"]]
        elif len(value) > rule["max"]:
            errors[name] = [f"Use até {rule['max']} caracteres."]
    if "email" in cleaned and not re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]+", cleaned["email"]):
        errors["email"] = ["Informe um e-mail válido."]
    if "email" in cleaned:
        cleaned["email"] = cleaned["email"].lower()
    identifier = payload.get("id")
    try:
        if not isinstance(identifier, str) or str(UUID(identifier)) != identifier.lower():
            raise ValueError
        cleaned["id"] = identifier.lower()
    except (ValueError, AttributeError):
        errors["id"] = ["Identificador de envio inválido."]
    for name, choices, message in [
        ("service", RULES["services"], "Selecione um serviço."),
        ("state", RULES["states"], "Selecione a UF."),
    ]:
        value = payload.get(name)
        if not isinstance(value, str) or value not in choices:
            errors[name] = [message]
        else:
            cleaned[name] = value
    cep = payload.get("cep")
    if not isinstance(cep, str) or not re.fullmatch(r"[0-9]{8}", cep):
        errors["cep"] = ["Informe um CEP com 8 dígitos."]
    else:
        cleaned["cep"] = cep
    if payload.get("consent") is not True:
        errors["consent"] = ["Autorize o uso dos dados para este contato."]
    cleaned["consent"] = 1
    if payload.get("website") != "":
        errors["website"] = ["Envio inválido."]
    return cleaned, errors

File: frontend/validation.js
import { z } from "zod";
import rules from "../shared/contact-rules.json" with { type: "json" };
export const { states, services } = rules;
const fields = Object.fromEntries(
  Object.entries(rules.fields).map(([name, rule]) => [
    name,
    z
      .string({
        required_error: rule.message,
        invalid_type_error: rule.message,
      })
      .trim()
      .min(rule.min, rule.message)
      .max(rule.max, `Use até ${rule.max} caracteres.`),
  ]),
);
export const contactSchema = z.object({
  ...fields,
  email: fields.email.regex(
    /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
    "Informe um e-mail válido.",
  ),
  id: z.string().uuid("Identificador de envio inválido."),
  service: z
    .string({ required_error: "Selecione um serviço." })
    .refine((v) => services.includes(v), "Selecione um serviço."),
  state: z
    .string({ required_error: "Selecione a UF." })
    .refine((v) => states.includes(v), "Selecione a UF."),
  cep: z.string().regex(/^[0-9]{8}$/, "Informe um CEP com 8 dígitos."),
  consent: z.literal(true, {
    errorMap: () => ({
      message: "Autorize o uso dos dados para este contato.",
    }),
  }),
  website: z.literal("", { errorMap: () => ({ message: "Envio inválido." }) }),
});
